api/models/source.go

```diff
@@ -0,0 +1,287 @@
+		HMac:      vc.HMac.transform(),
+		BasicAuth: vc.BasicAuth.transform(),
+		ApiKey:    vc.ApiKey.transform(),
+	}
+}
+
+type HMac struct {
+	Header   string                 `json:"header" valid:"required"`
+	Hash     string                 `json:"hash" valid:"supported_hash,required"`
+	Secret   string                 `json:"secret" valid:"required"`
+	Encoding datastore.EncodingType `json:"encoding" valid:"supported_encoding~please provide a valid encoding type,required"`
+}
+
+func (hm *HMac) transform() *datastore.HMac {
+	if hm == nil {
+		return nil
+	}
+
+	return &datastore.HMac{
+		Header:   hm.Header,
+		Hash:     hm.Hash,
+		Secret:   hm.Secret,
+		Encoding: hm.Encoding,
+	}
+}
+
+type BasicAuth struct {
+	UserName string `json:"username" valid:"required" `
+	Password string `json:"password" valid:"required"`
+}
+
+func (ba *BasicAuth) transform() *datastore.BasicAuth {
+	if ba == nil {
+		return nil
+	}
+
+	return &datastore.BasicAuth{
+		UserName: ba.UserName,
+		Password: ba.Password,
+	}
+}
+
+type ApiKey struct {
+	HeaderValue string `json:"header_value" valid:"required"`
+	HeaderName  string `json:"header_name" valid:"required"`
+}
+
+func (ak *ApiKey) transform() *datastore.ApiKey {
+	if ak == nil {
+		return nil
+	}
+
+	return &datastore.ApiKey{
+		HeaderValue: ak.HeaderValue,
+		HeaderName:  ak.HeaderName,
+	}
+}
+
+type PubSubConfig struct {
+	Type    datastore.PubSubType `json:"type"`
+	Workers int                  `json:"workers"`
+	Sqs     *SQSPubSubConfig     `json:"sqs"`
+	Google  *GooglePubSubConfig  `json:"google"`
+}
+
+func (pc *PubSubConfig) Transform() *datastore.PubSubConfig {
+	if pc == nil {
+		return nil
+	}
+
+	return &datastore.PubSubConfig{
+		Type:    pc.Type,
+		Workers: pc.Workers,
+		Sqs:     pc.Sqs.transform(),
+		Google:  pc.Google.transform(),
+	}
+}
+
+type SQSPubSubConfig struct {
+	AccessKeyID   string `json:"access_key_id"`
+	SecretKey     string `json:"secret_key"`
+	DefaultRegion string `json:"default_region"`
+	QueueName     string `json:"queue_name"`
+}
+
+func (sc *SQSPubSubConfig) transform() *datastore.SQSPubSubConfig {
+	if sc == nil {
+		return nil
+	}
+
+	return &datastore.SQSPubSubConfig{
+		AccessKeyID:   sc.AccessKeyID,
+		SecretKey:     sc.SecretKey,
+		DefaultRegion: sc.DefaultRegion,
+		QueueName:     sc.QueueName,
+	}
+}
+
+type GooglePubSubConfig struct {
+	SubscriptionID string `json:"subscription_id"`
+	ServiceAccount []byte `json:"service_account"`
+	ProjectID      string `json:"project_id"`
+}
+
+func (gc *GooglePubSubConfig) transform() *datastore.GooglePubSubConfig {
+	if gc == nil {
+		return nil
+	}
+
+	return &datastore.GooglePubSubConfig{
+		SubscriptionID: gc.SubscriptionID,
+		ServiceAccount: gc.ServiceAccount,
+		ProjectID:      gc.ProjectID,
+	}
+}
+
+type SourceResponse struct {
+	*datastore.Source
+}
+
+// Generic function for looping over a slice of type M
+// and returning a slice of type T
+func NewListResponse[T, M any](items []M, fn func(item M) T) []T {
```

Comment:
Should this function live here or in a general file?

Reply:
Moved it to `api/models.go`